File: smoke_tests/php/log_injection.php
<?php
// Log Injection vulnerabilities in PHP

// Test 1: error_log with user input
function log_error_unsafe() {
    $message = $_GET['message'];
    // VULNERABLE: User input can contain newlines
    error_log("[ERROR] User message: $message");
}

// Test 2: file_put_contents logging
function file_log() {
    $username = $_POST['username'];
    $action = $_POST['action'];
    // VULNERABLE: Can inject fake log entries
    $log = "[" . date('Y-m-d H:i:s') . "] User $username performed $action\n";
    file_put_contents('/var/log/app.log', $log, FILE_APPEND);
}

// Test 3: syslog with user input
function syslog_unsafe() {
    $user_input = $_GET['input'];
    openlog('myapp', LOG_PID, LOG_USER);
    // VULNERABLE: User input in syslog
    syslog(LOG_INFO, "User action: $user_input");
    closelog();
}

// Test 4: fwrite logging
function fwrite_log() {
    $event = $_POST['event'];
    $f = fopen('/var/log/events.log', 'a');
    // VULNERABLE: Event can contain CRLF
    fwrite($f, "[EVENT] $event\n");
    fclose($f);
}

// Test 5: JSON log breaking
function json_log() {
    $message = $_GET['msg'];
    // VULNERABLE: Can break JSON structure
    $log_entry = json_encode([
        'timestamp' => time(),
        'message' => $message
    ]);
    file_put_contents('/var/log/json.log', $log_entry . "\n", FILE_APPEND);
}

// Test 6: Monolog with user data
function monolog_log() {
    $data = $_POST['data'];
    $logger = new \Monolog\Logger('app');
    $logger->pushHandler(new \Monolog\Handler\StreamHandler('/var/log/app.log'));
    // VULNERABLE: User data in log context
    $logger->info("Processing: $data");
}

// Test 7: Apache error log
function apache_log() {
    $error = $_GET['error'];
    // VULNERABLE: User controls error message
    trigger_error("User error: $error", E_USER_WARNING);
}

// Test 8: Debug logging
function debug_log() {
    $debug_data = $_POST['debug'];
    // VULNERABLE: Debug data from user
    file_put_contents('/tmp/debug.log', print_r($debug_data, true), FILE_APPEND);
}

// Test 9: Exception message logging
function log_exception() {
    try {
        throw new Exception($_GET['error']);
    } catch (Exception $e) {
        // VULNERABLE: Exception message from user
        error_log("Exception: " . $e->getMessage());
    }
}

// Test 10: Audit log manipulation
function audit_log() {
    $user = $_POST['user'];
    $action = $_POST['action'];
    $result = $_POST['result'];
    // VULNERABLE: Multiple fields can contain newlines
    $audit = "User: $user\nAction: $action\nResult: $result\n---\n";
    file_put_contents('/var/log/audit.log', $audit, FILE_APPEND);
}

// Test 11: PSR-3 logger injection
function psr3_log() {
    $context = $_POST['context'];
    $logger = get_logger();
    // VULNERABLE: Context from user input
    $logger->info("Request processed", ['context' => $context]);
}

function get_logger() {
    // Return PSR-3 compatible logger
    return new class {
        public function info($message, $context = []) {
            error_log("$message " . json_encode($context));
        }
    };
}
?>

File: smoke_tests/ruby/log_injection.rb
# Log Injection vulnerabilities in Ruby
require 'logger'

class LogInjectionController < ApplicationController
  # Test 1: Rails logger with user input
  def log_action
    username = params[:username]
    # VULNERABLE: Username can contain newlines
    Rails.logger.info "User #{username} logged in"
  end

  # Test 2: Standard Logger
  def log_message
    message = params[:message]
    logger = Logger.new('/var/log/app.log')
    # VULNERABLE: Message not sanitized
    logger.info message
  end

  # Test 3: puts to log
  def debug_log
    data = params[:data]
    # VULNERABLE: Direct output to STDOUT
    puts "Processing: #{data}"
    head :ok
  end

  # Test 4: File.write logging
  def file_log
    event = params[:event]
    # VULNERABLE: Event can contain CRLF
    File.open('/var/log/events.log', 'a') do |f|
      f.puts "[#{Time.now}] #{event}"
    end
    head :ok
  end

  # Test 5: JSON log format breaking
  def json_log
    message = params[:message]
    # VULNERABLE: Can break JSON structure
    log_entry = { timestamp: Time.now, message: message }.to_json
    File.open('/var/log/json.log', 'a') { |f| f.puts log_entry }
    head :ok
  end

  # Test 6: Syslog injection
  def syslog_message
    user_input = params[:input]
    # VULNERABLE: User input in syslog
    Syslog.open('myapp') do |s|
      s.info user_input
    end
    head :ok
  end

  # Test 7: Exception message logging
  def log_exception
    begin
      raise params[:error]
    rescue => e
      # VULNERABLE: Exception message from user
      Rails.logger.error "Exception: #{e.message}"
    end
    head :ok
  end

  # Test 8: Audit log manipulation
  def audit_log
    user = params[:user]
    action = params[:action]
    result = params[:result]
    # VULNERABLE: Multiple fields can contain newlines
    audit = "User: #{user}\nAction: #{action}\nResult: #{result}"
    File.open('/var/log/audit.log', 'a') { |f| f.puts audit }
    head :ok
  end

  # Test 9: WARN level logging
  def warn_log
    warning = params[:warning]
    # VULNERABLE: Warning from user input
    Rails.logger.warn "Warning: #{warning}"
    head :ok
  end

  # Test 10: Debug logging in production
  def debug_info
    debug_data = params[:debug]
    # VULNERABLE: Debug info in production
    Rails.logger.debug "Debug: #{debug_data}"
    head :ok
  end

  # Test 11: Tagged logging
  def tagged_log
    tag = params[:tag]
    message = params[:message]
    # VULNERABLE: Both tag and message from user
    Rails.logger.tagged(tag) do
      Rails.logger.info message
    end
    head :ok
  end
end
